File: bridge/voice.js
"use strict";

const WRAPPER_KEYS = [
  "ephemeralMessage",
  "viewOnceMessage",
  "viewOnceMessageV2",
  "viewOnceMessageV2Extension",
  "documentWithCaptionMessage",
];
const MAX_AUDIO_BYTES = 8 * 1024 * 1024;

function findAudioMessage(content, depth = 0) {
  if (!content || depth > 5 || typeof content !== "object") return null;
  if (content.audioMessage && typeof content.audioMessage === "object") {
    return content.audioMessage;
  }
  for (const key of WRAPPER_KEYS) {
    const wrapped = content[key]?.message;
    const audio = findAudioMessage(wrapped, depth + 1);
    if (audio) return audio;
  }
  return null;
}

function extensionForMime(mimeType) {
  const normalized = String(mimeType || "").split(";", 1)[0].trim().toLowerCase();
  return {
    "audio/ogg": "ogg",
    "audio/mpeg": "mp3",
    "audio/mp4": "m4a",
    "audio/wav": "wav",
    "audio/webm": "webm",
  }[normalized] || "ogg";
}

async function transcribeIncomingAudio({
  message,
  downloadMediaMessage,
  reuploadRequest,
  transcribeUrl,
  authToken,
  fetchImpl = fetch,
}) {
  const content = message?.message || message;
  const audioMessage = findAudioMessage(content);
  if (!audioMessage) return null;
  const advertisedSize = Number(audioMessage.fileLength || 0);
  if (Number.isFinite(advertisedSize) && advertisedSize > MAX_AUDIO_BYTES) {
    throw new Error("Ukuran voice note terlalu besar untuk diproses.");
  }
  if (!authToken) throw new Error("Token transkripsi bridge belum dikonfigurasi.");

  const audioBuffer = await downloadMediaMessage(
    message,
    "buffer",
    {},
    { reuploadRequest },
  );
  const response = await fetchImpl(transcribeUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      "X-RESTO-API-TOKEN": authToken,
    },
    body: JSON.stringify({
      audio_base64: Buffer.from(audioBuffer).toString("base64"),
      mime_type: audioMessage.mimetype || "audio/ogg",
      filename: `voice-note.${extensionForMime(audioMessage.mimetype)}`,
    }),
  });
  const payload = await response.json();
  if (!response.ok) {
    throw new Error(String(payload?.detail || "Voice note belum dapat diproses."));
  }
  const transcript = String(payload?.transcript || "").trim();
  if (!transcript) throw new Error("Isi voice note belum dapat dikenali.");
  return transcript;
}

module.exports = { findAudioMessage, transcribeIncomingAudio };

File: app/voice.py
"""OpenAI-backed speech-to-text helper for inbound WhatsApp voice notes."""

from __future__ import annotations

import json
import os
import uuid
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen


class VoiceTranscriptionError(RuntimeError):
    """Raised when a voice note cannot be transcribed."""


MAX_AUDIO_BYTES = 8 * 1024 * 1024
SUPPORTED_AUDIO_MIME_TYPES = {
    "audio/ogg",
    "audio/mpeg",
    "audio/mp4",
    "audio/wav",
    "audio/webm",
}


def validate_voice_input(audio_bytes: bytes, mime_type: str) -> str:
    """Validate bounded WhatsApp audio before invoking the paid provider."""
    if not audio_bytes:
        raise VoiceTranscriptionError("Voice note kosong atau tidak dapat diunduh.")
    if len(audio_bytes) > MAX_AUDIO_BYTES:
        raise VoiceTranscriptionError("Ukuran voice note terlalu besar untuk diproses.")
    normalized_mime = mime_type.split(";", 1)[0].strip().lower()
    if normalized_mime not in SUPPORTED_AUDIO_MIME_TYPES:
        raise VoiceTranscriptionError("Format audio belum didukung.")
    return normalized_mime


def _multipart_body(
    *,
    model: str,
    audio_bytes: bytes,
    mime_type: str,
    filename: str,
) -> tuple[bytes, str]:
    boundary = f"----resto-ai-{uuid.uuid4().hex}"
    line = b"\r\n"
    normalized_mime = mime_type.split(";", 1)[0].strip().lower() or "application/octet-stream"
    body = b"".join(
        [
            f"--{boundary}".encode(),
            line,
            b'Content-Disposition: form-data; name="model"',
            line,
            line,
            model.encode("utf-8"),
            line,
            f"--{boundary}".encode(),
            line,
            (
                'Content-Disposition: form-data; name="file"; '
                f'filename="{filename}"'
            ).encode("utf-8"),
            line,
            f"Content-Type: {normalized_mime}".encode("utf-8"),
            line,
            line,
            audio_bytes,
            line,
            f"--{boundary}--".encode(),
            line,
        ]
    )
    return body, boundary


def transcribe_audio(
    audio_bytes: bytes,
    mime_type: str,
    *,
    filename: str = "voice-note.ogg",
) -> str:
    """Send one WhatsApp voice note to OpenAI and return its transcript."""
    api_key = os.getenv("OPENAI_API_KEY", "").strip()
    if not api_key:
        raise VoiceTranscriptionError("Layanan transkripsi belum dikonfigurasi.")

    normalized_mime = validate_voice_input(audio_bytes, mime_type)
    model = os.getenv("OPENAI_TRANSCRIPTION_MODEL", "gpt-4o-mini-transcribe").strip() or "gpt-4o-mini-transcribe"
    body, boundary = _multipart_body(
        model=model,
        audio_bytes=audio_bytes,
        mime_type=normalized_mime,
        filename=filename,
    )
    request = Request(
        "https://api.openai.com/v1/audio/transcriptions",
        data=body,
        headers={
            "Authorization": f"Bearer {api_key}",
            "Content-Type": f"multipart/form-data; boundary={boundary}",
        },
        method="POST",
    )
    try:
        with urlopen(request, timeout=30) as response:
            payload = json.loads(response.read().decode("utf-8"))
    except (HTTPError, URLError, TimeoutError, ValueError) as exc:
        raise VoiceTranscriptionError("Voice note belum dapat ditranskripsikan.") from exc

    transcript = str(payload.get("text", "")).strip()
    if not transcript:
        raise VoiceTranscriptionError("Isi voice note belum dapat dikenali.")
    return transcript
